### DISCUSSION TITLE
Table interface for tabular data

### DISCUSSION THREAD

**Discussion by stx-chris** [directus/directus#14847] [Feature Requests]
We've had some use cases where we had to store certain data in a tabular fashion. This is often the case when data is exchanged with the client through Excel. It is cumbersome to split the data into its relations and import it into Directus. Dividing the data into individual fields is also no option. We also tried the Repeater Interface, but to mass-edit entries one had to open each individual record, save the changes and proceed with the next.

We would rather have a tabular interface where the data, stored as JSON, can be mass edited and filtered later on in the consuming app. 

With headers being configurable, the interface could allow for entries such as:

```
name       | result      | category | value  | comment
-----------|-------------|----------|--------|------------------
result1    | 1st result  | high     | 123.10 | important for xyz
-----------|-------------|----------|--------|------------------
result2    | 2nd result  | medium   | 413.10 | 
-----------|-------------|----------|--------|------------------
result3    | 3rd result  | low      |   3.20 | can be ignored
```

The data could then be stored as JSON as follows, possibly allowing for HTML as well:
```
{
   "header":["name", "result", "category", "value", "comment"],
   "rows":[{
         "name": { "type": ..., "label": "result1", ... }
         "result": { "type": ..., "label": "1st result", ... }
         "comment": { "type": ..., "label": "<b>important</b> for xyz", ... }
      }, {
         "name": { "type": ..., "label": "result2", ... }
         "result": { "type": ..., "label": "2nd result", ... }
      }...
   ]
}
```

The naming needs to be adjusted to the current way how Directus stores field metadata.

**Comment by rijkvanzanten** [comment]
So would you envision this to be an editing experience like an inline spreadsheet of sorts? To that extend, do you expect the values as nested JSON objects like you showed, or would a CSV type value make more sense for the type of data? Like

```
[
  ['name',    'result',     'category', 'value', 'comment'                  ],
  ['result1', '1st result', 'high',     123.10,  '<b>important</b> for xyz' ],
  ['result2', '2nd result', 'medium',   413.10,  null                       ],
  ['result3', '3rd result', 'low',      3.20,    'can be ignored'           ]
]
```

**Reply by stx-chris** [reply]
Exactly @rijkvanzanten, the goal is to achieve a spreadsheet-like editing experience. A csv/array type of table would definitely be preferable to JSON, also much easier to import and export to spreadsheet applications. The question is, though, how different column data types can be specified and stored. Maybe it's something that can be configured in the Interface settings. 

Even if every cell is stored as a string and cast later in the consuming app, it would save a lot of time entering and maintaining the values.

**Reply by rijkvanzanten** [reply]
> Maybe it's something that can be configured in the Interface settings.

Totally! It sounds very similar to the "repeater" interface to me, but instead of repeating rows with a full form that opens up for the editing experience, everything is inline. 

Do you imagine that each "cell" renders an interface still, or would it be "raw values" everywhere?

**Reply by stx-chris** [reply]
I had another look at the repeater interface, which stores its data as JSON. On second thought, if someone needs to deal with different cell data types, the repeater is the way to go. For the table interface I see no reason for each individual cell to have its own interface. It would suffice to have equally split column widths and auto-expanding row heights for raw value entry, similar to what Atlassian's Confluence offers. 

The data can be stored as a nested array of strings, as proposed by you. If a data type cast (e.g. from string to double) is necessary, the consuming app should deal with it, just like regular CSV processing.

e.g. [Table interface](https://confluence.atlassian.com/doc/tables-136463.html) in Confluence:
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/62717406/183235595-85bc1dfd-526b-417f-96bf-e7dee34114a2.png">

**Comment by loongmxbt** [comment]
@stx-chris Please look at this extension: https://github.com/seymoe/directus-extension-vgrid-interface
This supports various json tabular data format and also csv.

**Reply by stx-chris** [reply]
thx @loongmxbt this is awesome! @rijkvanzanten any chance such a data table interface (ie. a more robust version of it and supporting more options, such as a different delimiter) might become a standard interface in Directus?

@rijkvanzanten Following up on our earlier conversation, this implementation uses csv-format directly when encoding the table contents, e.g.

```
name,de-DE,en-US
firstname,Vorname,"First name, or Given name"
lastname,Nachname,Last name
address,Adresse,Address
```
is stored as a single string which makes importing and exporting much easier than an array structure. For handling the data in a consuming app, a parser would come in handy, though.

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten any chance such a data table interface (ie. a more robust version of it and supporting more options, such as a different delimiter) might become a standard interface in Directus?

Could be! It's always a fine balance between maintenance overhead and usefulness for "the greater good". We normally go by our 80/20 rule, which effectively states that we only implement the things in core that are reasonably useful for 80+% of users, and keep the rest available through extensions. As of right now, this feels like more of an extension to me than a core interface (because of its specific use case and proprietary data format), but lets see how much engagement we get on this feature request. I might be mistaken 🙂 

We're also looking long and hard at https://github.com/directus/directus/discussions/2820 to make that type of extension distribution _way_ easier, which should really help the discovery and "installability" of these types of extensions 🙂

**Reply by LasseRosenow** [reply]
I also really would love to have some table interface. Many customers that we create websites for, want to have tables in their content and right now what we do is use a wysiwyg editor and create the textfield in there. But we keep on having to remind them how it works. 
They always forget it and wonder where would this table content element be?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
